File: server/backend/src/backend/api/models/puller.py
from __future__ import annotations

from typing import Literal

from backend.api.models.base import ResponseBase
from backend.api.models.operations import TaskOperationRead


class FrontendPullData(ResponseBase):
    """Live execution state delivered through the single frontend polling pipeline."""

    operations: list[TaskOperationRead]


class FrontendPullAPIRead(ResponseBase):
    """One generic execution snapshot and the cadence the frontend should use next."""

    mode: Literal["slow", "fast"]
    data: FrontendPullData

File: ui/src/types/schemas/puller.ts
import {z} from 'zod'
import {TaskOperationReadSchema} from './operation'


export const FrontendPullDataSchema = z.object({
  operations: TaskOperationReadSchema.array(),
})

export const FrontendPullReadSchema = z.object({
  mode: z.enum(['slow', 'fast']),
  data: FrontendPullDataSchema,
})

export type FrontendPullData = z.infer<typeof FrontendPullDataSchema>
export type FrontendPullRead = z.infer<typeof FrontendPullReadSchema>
